processor/transformprocessor/internal/metrics/func_drop_bucket.go

```diff
@@ -0,0 +1,133 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package metrics // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor/internal/metrics"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlmetric"
+)
+
+type DropBucketArguments struct {
+	Pattern string
+}
+
+func newDropBucketFactory() ottl.Factory[ottlmetric.TransformContext] {
+	return ottl.NewFactory(
+		"drop_bucket",
+		&DropBucketArguments{},
+		createDropBucketFunction,
+	)
+}
+
+func createDropBucketFunction(_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[ottlmetric.TransformContext], error) {
+	args, ok := oArgs.(*DropBucketArguments)
+
+	if !ok {
+		return nil, errors.New("DropBucketFactory args must be of type *DropBucketArguments[K]")
+	}
+
+	return dropBucketFunc(args.Pattern), nil
+}
+
+func dropBucketFunc(pattern string) ottl.ExprFunc[ottlmetric.TransformContext] {
+	return func(_ context.Context, tCtx ottlmetric.TransformContext) (any, error) {
+		compiledPattern, err := regexp.Compile(pattern)
+		if err != nil {
+			return nil, fmt.Errorf("the regex pattern supplied to drop_bucket is not a valid pattern: %w", err)
+		}
+
+		metric := tCtx.GetMetric()
+
+		switch metric.Type() {
+		case pmetric.MetricTypeHistogram:
+			dropBucketHistogram(&metric, compiledPattern)
+		case pmetric.MetricTypeGauge:
+			dropBucketGauge(&metric, compiledPattern)
+		case pmetric.MetricTypeSum:
+			dropBucketSum(&metric, compiledPattern)
+		}
+		return nil, nil
+	}
+}
+
+func dropBucketHistogram(metric *pmetric.Metric, compiledPattern *regexp.Regexp) {
```

Comment:
Logic here breaks a very important property of the histograms, see https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L446 and https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L456

Reply:
Thanks for pointing out this. I will fix it.